hydra-node/src/Hydra/Logging/Monitoring.hs

```diff
@@ -38,19 +43,55 @@ withMonitoring (Just monitoringPort) (Tracer tracer) action = do
 -- | Register all relevant metrics.
 -- Returns an updated `Registry` which is needed to `serveMetrics` or any other form of publication
 -- of metrics, whether push or pull, and a function for updating metrics given some trace event.
-prepareRegistry :: MonadIO m => m (HydraLog tx net -> m (), Registry)
-prepareRegistry = first monitor <$> registerMetrics
+prepareRegistry :: (MonadIO m, MonadSTM m, MonadMonotonicTime m, Tx tx) => TVar m (Map (TxId tx) Time) -> m (HydraLog tx net -> m (), Registry)
+prepareRegistry txMap = first monitor <$> registerMetrics
  where
+  monitor metricsMap (Node (ProcessingEvent _ (NetworkEvent (ReqTx _ tx)))) = do
+    t <- getMonotonicTime
+    atomically $ modifyTVar' txMap (Map.insert (txId tx) t)
+    tick "hydra_head_requested_tx" metricsMap
+  monitor metricsMap (Node (ProcessedEffect _ (ClientEffect (SnapshotConfirmed snapshot)))) = do
+    t <- getMonotonicTime
+    forM_ (confirmed snapshot) $ \tx -> do
+      let tid = txId tx
+      txsStartTime <- readTVarIO txMap
+      case Map.lookup tid txsStartTime of
+        Just start -> histo "hydra_head_tx_confirmation_time_ms" (diffTime t start) metricsMap
+        Nothing -> pure ()
+    tickN (length $ confirmed snapshot) "hydra_head_confirmed_tx" metricsMap
   monitor metricsMap (Node (ProcessedEvent _ _)) =
-    case Map.lookup "hydra_head_events" metricsMap of
+    tick "hydra_head_events" metricsMap
+  monitor _ _ = pure ()
+
+  tick metricName metricsMap =
+    case Map.lookup metricName metricsMap of
       (Just (CounterMetric c)) -> liftIO $ inc c
       _ -> pure ()
-  monitor _ _ = pure ()
+
+  tickN num metricName metricsMap =
+    case Map.lookup metricName metricsMap of
+      (Just (CounterMetric c)) -> liftIO $ add num c
+      _ -> pure ()
+
+  histo metricName time metricsMap =
+    case Map.lookup metricName metricsMap of
+      (Just (HistogramMetric h)) -> liftIO $ observe (fromRational $ toRational $ time * 1000) h
+      _ -> pure ()
 
   registerMetrics = foldlM registerMetric (mempty, new) allMetrics
 
-  allMetrics = [(Name "hydra_head_events", CounterMetric, flip registerCounter mempty)]
+  allMetrics :: [MetricDefinition]
+  allMetrics =
+    [ MetricDefinition (Name "hydra_head_events") CounterMetric $ flip registerCounter mempty
+    , MetricDefinition (Name "hydra_head_requested_tx") CounterMetric $ flip registerCounter mempty
+    , MetricDefinition (Name "hydra_head_confirmed_tx") CounterMetric $ flip registerCounter mempty
+    , MetricDefinition (Name "hydra_head_tx_confirmation_time_ms") HistogramMetric $ \n -> registerHistogram n mempty [100, 200 .. 2000]
+    ]
```

Comment:
The `MetricDefinition` does not add much value IMO, but your obviously free to keep it :)

Reply:
It's an existential type so the constructor is needed: previous version simply used tuples but it failed when I started adding histogram